### PR TITLE
[SPARK-49355][SQL] `levenshtein` should check whether the `collation` values of all parameter types are the same

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The same principle as https://github.com/apache/spark/pull/47825#pullrequestreview-2250729020, the parameters (`left` and `right`) in expression `levenshtein` are `collation-dependent`, rather than `collation-unaware`.

### Why are the changes needed?
Strengthen the parameter data type check of expression `levenshtein`  to avoid potential issues.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Add some `test case` to `collations.sql`.
- Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala [hunk 1]
```diff
@@ -90,6 +90,15 @@ object CollationTypeCasts extends TypeCoercionRule {
       val newValues = collateToSingleType(mapCreate.values)
       mapCreate.withNewChildren(newKeys.zip(newValues).flatMap(pair => Seq(pair._1, pair._2)))
 
+    case levenshtein: Levenshtein =>
+      val Seq(left, right) = Seq(levenshtein.children.head, levenshtein.children(1))
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala:94]:
```suggestion
      val Seq(left, right, threshold) = levenshtein.children
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala [hunk 2]
```diff
@@ -90,6 +90,15 @@ object CollationTypeCasts extends TypeCoercionRule {
...
+      val Seq(newLeft, newRight) = collateToSingleType(Seq(left, right))
+      if (levenshtein.children.size == 2) {
+        levenshtein.withNewChildren(Seq(newLeft, newRight))
+      } else {
+        levenshtein.withNewChildren(Seq(newLeft, newRight, levenshtein.children(2)))
```
[the remark this answers is not in the document]
Comment (uros-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala:99]: or even maybe try to combine the if/else branches using something like:
`levenshtein.withNewChildren(Seq(newLeft, newRight) ++ rest)`

[the remark this answers is not in the document]
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala]: Updated, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala (hunk 1 above)
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala]: Very concise writing, thank you!

**Review by uros-db — APPROVED** [review]
lgtm

**Comment by panbingkun** [comment]
cc @MaxGekk @cloud-fan

**Review by MaxGekk — COMMENTED** [review]
@panbingkun Could you resolve conflicts, please.

**Comment by panbingkun** [comment]
> @panbingkun Could you resolve conflicts, please.

Done, thanks!

**Comment by MaxGekk** [comment]
@panbingkun The failed GA [Run / Linters, licenses, and dependencies](https://github.com/panbingkun/spark/actions/runs/10560715820/job/29254910061#logs) is not related to your changes, I think, but could you re-run this GA only just in case.

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun and @uros-db @viktorluc-db for review.

**Pull request closed without merging**